src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -47,6 +50,20 @@ public KafkaProducerMessageHandler(final KafkaProducerContext kafkaProducerConte
 		this.kafkaProducerContext = kafkaProducerContext;
 	}
 
+	/**
+	 * Enables the use of headers for determining the target topic and partition of outbound messages. By default it is
+	 * set to true, but it can be disabled when those values are produced by upstream components that read messages
+	 * from Kafka sources themselves.
+	 *
+	 * @see KafkaHeaders#TOPIC
+	 * @see KafkaHeaders#PARTITION_ID
+	 * @param enableHeaderRouting whether the topic and destination headers should be considered
```

Comment:
Yes, as a gold middle we can live with this idea as well. At least we can simply it include to the `1.3` and silently fix the XD issue just only with one more option.
We should add a note somewhere or `TODO` on those headers resolution.
And really remove this logic in the `2.0` in favor of the provided `expression`. Thus we are washing our hands and go to drink beer :smile: